src/App.tsx

```diff
@@ -0,0 +1,142 @@
+
+const STORAGE_KEY = 'vanilla_todo_step3' // 바닐라와 동일 키
+
+const AppCard = styled.main`
+  background: #fdf6f0;
+  padding: 20px;
+  border-radius: 16px;
+  box-shadow: 0 6px 20px rgba(0, 0, 0, 0.15);
+  width: 300px;
+  display: flex;
+  flex-direction: column;
+  gap: 16px;
+`
+
+const Header = styled.header`
+  display: flex;
+  flex-direction: column;
+  gap: 10px;
+  align-items: center;
+
+  h1 {
+    margin: 0;
+    text-align: center;
+  }
+`
+
+const ListSection = styled.section<{ scrolling: boolean; maxH: number | null }>`
+  ${({ scrolling, maxH }) => (scrolling ? `overflow-y:auto; max-height:${maxH ?? 0}px;` : '')}
+`
+
+function App() {
+  const [todos, setTodos] = useState<Todo[]>([])
+  const [currentDate, setCurrentDate] = useState<string>(() => toYYYYMMDD(new Date()))
+
+  // 스크롤 영역 계산용
+  const listRef = useRef<HTMLUListElement | null>(null)
+  const [scrolling, setScrolling] = useState(false)
+  const [maxH, setMaxH] = useState<number | null>(null)
+
+  // 초기 로드
+  useEffect(() => {
+    const raw = localStorage.getItem(STORAGE_KEY)
+    if (raw) {
+      const parsed = JSON.parse(raw) as Todo[]
+      // 과거 데이터에 due 없으면 현재 날짜로 보정
+      const fixed = parsed.map((t) => (t.due ? t : { ...t, due: currentDate }))
+      setTodos(fixed)
+    }
+  }, []) // eslint-disable-line
+
+  // 저장
+  useEffect(() => {
+    localStorage.setItem(STORAGE_KEY, JSON.stringify(todos))
+  }, [todos])
+
+  // 현재 날짜의 목록
+  const todays = useMemo(() => todos.filter((t) => t.due === currentDate), [todos, currentDate])
+
+  // 5개 초과 시 높이 동적 계산(바닐라와 동일한 로직)
+  useEffect(() => {
+    const ul = listRef.current
+    if (!ul) return
+
+    const items = Array.from(ul.querySelectorAll<HTMLLIElement>('.todo-item'))
+    if (items.length > 5) {
+      let max = 0
+      for (let i = 0; i < 5; i++) {
+        max += items[i].getBoundingClientRect().height
+      }
+      const cs = getComputedStyle(ul)
+      const rowGap = parseFloat(cs.rowGap || (cs as any).gap || '0')
+      max += rowGap * 4
+      setMaxH(Math.ceil(max))
+      setScrolling(true)
+    } else {
+      setMaxH(null)
+      setScrolling(false)
+    }
+  }, [todays.length])
+
+  // 추가/토글/삭제
+  const addTodo = (text: string) => {
+    setTodos((prev) => [
+      ...prev,
+      {
+        id: `${Date.now()}${Math.random().toString(16).slice(2)}`,
+        text,
+        done: false,
+        due: currentDate,
+      },
+    ])
+  }
+  const toggleTodo = (id: string) => {
+    setTodos((prev) => prev.map((t) => (t.id === id ? { ...t, done: !t.done } : t)))
+  }
+  const removeTodo = (id: string) => {
+    setTodos((prev) => prev.filter((t) => t.id !== id))
+  }
+
+  // 날짜 이동/선택
+  const goPrev = () => {
+    const d = new Date(currentDate + 'T00:00:00')
+    d.setDate(d.getDate() - 1)
+    setCurrentDate(toYYYYMMDD(d))
+  }
+  const goNext = () => {
+    const d = new Date(currentDate + 'T00:00:00')
+    d.setDate(d.getDate() + 1)
+    setCurrentDate(toYYYYMMDD(d))
+  }
```

Comment:
지금 goPrev와 goNext가 -1, +1 숫자만 다르고 로직은 동일한 상황이니, 아래와 같이 중복을 줄이는 방향으로 바꾸는 방법도 있을 것 같네요!

```suggestion
  // 날짜 이동/선택
  const changeDate = (offset: number) => {
  const d = new Date(currentDate + 'T00:00:00')
  d.setDate(d.getDate() + offset)
  setCurrentDate(toYYYYMMDD(d))
}

const goPrev = () => changeDate(-1)
const goNext = () => changeDate(1)
```